File: local_operator/mobile/web/src/themes/rose-pine.ts
import type { ThemeDefinition } from "./palette-contract";

/**
 * Rosé Pine, all three published variants.
 *
 * Upstream: github.com/rose-pine/palette, the canonical hex list in
 * `dist/css/rose-pine.css`. Each variant ships nine neutral roles
 * (`base surface overlay muted subtle text highlight-low/med/high`) and six
 * accents (`love gold rose pine foam iris`), against this contract's
 * twenty-nine. The neutrals map almost verbatim; the accents map by
 * upstream's own usage table rather than by hue:
 *
 * - `love` is the documented terminal red and error colour -> `danger`.
 * - `gold` is the documented warnings colour -> `warning`.
 * - `foam` is the documented "object keys, info, git add" hue -> `info`.
 * - `pine` is the documented terminal green (functions) -> `success`. It is
 *   a blue-green rather than a leaf green, which is Rosé Pine's signature
 *   inversion and is kept.
 * - `rose` takes `accent`: the theme's own name, and upstream's "git change,
 *   git dirty" hue - the colour of something happening.
 *
 * `iris` has no role in this contract (the TUI spends it on metadata labels,
 * which this app has no equivalent of), so it appears only in the hover and
 * active steps where the ramp needs a neighbour.
 *
 * ## What the four grounds cost
 *
 * Rosé Pine defines three grounds and this contract needs four. Each variant
 * therefore derives one step, and - importantly - the contract measures every
 * ink and semantic against ALL FOUR, including the lightest. That is a
 * strictly harder test than the ratios Rosé Pine's own docs quote, which is
 * why several accents that look canonical elsewhere are lifted here: the
 * binding ground is `elevated`, not `canvas`. Every deviation below carries
 * the canonical value and the measured reason.
 *
 * @see docs/branding.md § 3 - the floors these values satisfy
 */

/** Rosé Pine (main): the original night. */
export const rosePine: ThemeDefinition = {
	id: "rosePine",
	name: "Rosé Pine",
	description: "All natural pine, faux fur and a bit of soho vibes.",
	palette: {
		mode: "dark",

		canvas: "#191724",
		surface: "#1f1d2e",
		elevated: "#26233a",
		// Upstream has no ground below base. One step down, far enough to stay
		// visible at this darkness (a one-level step measures under the
		// separation floor down here).
		sunken: "#12101b",

		ink: "#e0def4",
		// Upstream's `subtle` #908caa clears the 4.5 floor on all four grounds
		// (5.48 / 5.12 / 4.71 / 5.84), so this pair is a HEADROOM choice, not a
		// floor rescue: on `elevated` it clears by 0.21, and this contract has
		// two mid inks where Rose Pine has one, so `inkDim` must sit BELOW
		// `inkMuted` and would inherit that thin margin. Both are lifted along
		// the same violet-gray toward `text` to buy the second rung room
		// (`inkDim` #a9a5c4 lands at 6.41 on `elevated`). Canonical `muted`
		// #6e6a86 is `inkDisabled`, which is what upstream uses it for
		// (comments, git-ignored) and the one role exempt from the floors.
		inkMuted: "#c5c2dd",
		inkDim: "#a9a5c4",
		inkDisabled: "#6e6a86",

		// Canonical highlight-low: upstream's cursor-line colour, which is exactly
		// a decorative rule's job.
		hairline: "#21202e",
		// Canonical highlight-high #524f67 is 2.25:1 on canvas - it is a highlight,
		// not a boundary, and every input in the app would have been bounded under
		// the 3:1 structural floor. Lifted along the same hue to clear that floor
		// on all four grounds (5.12:1 / 4.40:1 at the extremes).
		borderControl: "#8b87a3",

		accent: "#ebbcba",
		accentHover: "#f2cfcd",
		accentActive: "#d9a5a3",
		accentWash: "#2e2430",
		// Rose is a pale pink at 10.45:1 on canvas, so ink on the accent fill is
		// the page ground rather than a light value.
		onAccent: "#191724",

		// Canonical pine #31748f measures 3.38:1 on canvas and 2.91:1 on
		// `elevated` - the only canonical accent under the floor here, and under
		// it on every ground. Lifted along its own blue-green to 5.35:1 / 4.60:1.
		success: "#5995b2",
		successWash: "#1a2730",
		successBorder: "#417d99",

		warning: "#f6c177",
		warningWash: "#2b2419",
		warningBorder: "#8a7040",

		danger: "#eb6f92",
		dangerWash: "#2e1c26",
		dangerBorder: "#9e5468",

		info: "#9ccfd8",
		infoWash: "#1a2a2e",
		infoBorder: "#4f8189",

		overlayShadow: "0 12px 32px -12px rgb(14 12 20 / 0.7)",
		scrim: "rgb(14 12 20 / 0.6)",
	},
};

/** Rosé Pine Moon: the same palette one shade warmer and lighter. */
export const rosePineMoon: ThemeDefinition = {
	id: "rosePineMoon",
	name: "Rosé Pine Moon",
	description: "Rosé Pine's warmer moonlit ground, one shade up.",
	palette: {
		mode: "dark",

		canvas: "#232136",
		surface: "#2a273f",
		elevated: "#393552",
		sunken: "#1b192a",

		ink: "#e0def4",
		inkMuted: "#c9c6e0",
		inkDim: "#b0acc9",
		inkDisabled: "#6e6a86",

		hairline: "#2a283e",
		borderControl: "#8f8ba9",

		// Canonical moon rose #ea9a97 sits 43 RGB units from love #eb6f92 - on
		// main they are 87 apart - close enough that the primary action and an
		// error message read as the same hue. Desaturated one step to #eaaca9,
		// still moon's rose, now 32 ΔE from danger.
		accent: "#eaaca9",
		accentHover: "#f2c2bf",
		accentActive: "#d69793",
		accentWash: "#382b38",
		onAccent: "#232136",

		// Canonical moon pine #3e8fb0 is 3.19:1 on `elevated`. Lifted, and kept
		// 24.5 ΔE off foam so success and info stay two signals.
		success: "#4eadd9",
		successWash: "#22303c",
		successBorder: "#4a8aa8",

		warning: "#f6c177",
		warningWash: "#37301f",
		warningBorder: "#8f7546",

		// Canonical love #eb6f92 is 4.00:1 on moon's lighter `elevated` - the same
		// red that clears comfortably on main. Lifted 4.4 ΔE to 4.60:1.
		danger: "#f97a9e",
		dangerWash: "#38222e",
		dangerBorder: "#a35a72",

		info: "#9ccfd8",
		infoWash: "#23343a",
		infoBorder: "#57868f",

		overlayShadow: "0 12px 32px -12px rgb(20 18 32 / 0.7)",
		scrim: "rgb(20 18 32 / 0.6)",
	},
};

/** Rosé Pine Dawn: the light variant. */
export const rosePineDawn: ThemeDefinition = {
	id: "rosePineDawn",
	name: "Rosé Pine Dawn",
	description: "Rosé Pine at sunrise, warm paper under a soho morning.",
	palette: {
		mode: "light",

		/*
		 * The ground ladder inverts, and Rosé Pine hands it over almost intact:
		 * canonical dawn `surface` #fffaf3 is LIGHTER than `base` #faf4ed, so
		 * `base` is the page ground, canonical `surface` is the raised one, and
		 * canonical `highlight-low` #f4ede8 - which upstream paints the cursor
		 * line with - becomes the recessed ground. Only `elevated` is derived,
		 * warmed rather than merely lightened for the reason the brand light
		 * ramp documents: a near-white ladder has no lightness headroom left, so
		 * the steps separate on the chroma axis. Walking the ladder from `sunken`
		 * up, adjacent steps measure ΔE 2.06 / 2.04 / 2.55 - just past the ~2 the
		 * eye needs.
		 */
		canvas: "#faf4ed",
		surface: "#fffaf3",
		elevated: "#fffdf9",
		sunken: "#f4ede8",

		// Canonical text #575279 is 6.66:1 on `base` and 6.29:1 on the recessed
		// ground, under the 7:1 primary floor on both. Deepened 3.8 ΔE.
		ink: "#4e4970",
		inkMuted: "#6d6785",
		// Canonical muted #9893a5 is 2.73:1 - a comment colour, not a caption
		// colour. It moves to `inkDisabled` (the floor-exempt role it belongs in)
		// and the tertiary ink is a darker neutral on the same violet-gray.
		inkDim: "#6c697a",
		inkDisabled: "#9893a5",

		hairline: "#e6ddd6",
		// Canonical highlight-high #cecacd is 1.48:1: a highlight, not a boundary.
		borderControl: "#91838c",

		/*
		 * Every canonical dawn accent except pine misses the 4.5:1 floor on the
		 * four grounds - gold 2.05:1, rose 2.60:1, foam 3.14:1, iris 3.47:1,
		 * love 3.84:1 - so each is darkened along its own hue by the smallest
		 * step that clears it.
		 *
		 * `accent` (rose) and `danger` (love) both converge toward the same dusty
		 * red as they darken; at 15.3 ΔE apart they are the tightest pair here,
		 * which is deliberate - holding them further apart meant abandoning one
		 * of the two canonical hues entirely.
		 */
		accent: "#a15552", // rose #d7827e: 2.60:1
		accentHover: "#8b4644",
		accentActive: "#743a38",
		accentWash: "#fbece8",
		onAccent: "#fffaf3",

		success: "#286983", // pine - CANONICAL, the only accent that clears unaided
		successWash: "#e4edf1",
		successBorder: "#5c8ba0",

		warning: "#995d00", // gold #ea9d34: 2.05:1 - the largest deviation
		warningWash: "#f6ead6",
		warningBorder: "#b08b4c",

		danger: "#a1526a", // love #b4637a: 3.84:1
		dangerWash: "#fcebef",
		dangerBorder: "#b57e90",

		// foam #56949f: 3.14:1, and darkening it along its own hue walks it
		// straight into pine. Pushed to the teal side so info and success stay
		// 20.1 ΔE apart, which is the separation the other palettes here hold.
		info: "#107574",
		infoWash: "#e2eeef",
		infoBorder: "#5b979a",

		overlayShadow: "0 12px 32px -12px rgb(87 82 121 / 0.24)",
		scrim: "rgb(87 82 121 / 0.4)",
	},
};

File: local_operator/tui/palettes/rose_pine.py
"""Rosé Pine — all three published variants (main, moon, dawn).

Upstream: github.com/rose-pine/palette, the canonical hex list in
``dist/css/rose-pine.css``. Rosé Pine ships fifteen roles per variant — nine
neutral (``base surface overlay muted subtle text highlight-low/med/high``)
and six accent (``love gold rose pine foam iris``) — against this UI's
twenty-three semantic tokens, so the port is a MAPPING, and the mapping is
the interesting part of this file.

The neutral roles land almost verbatim. Rosé Pine's ``base`` is ``bg``,
``surface`` is ``surface``, ``overlay`` is ``raised``; the fourth elevation
step this UI needs has no upstream value, so ``overlay`` is interpolated one
just-visible step past the canonical ``overlay``, in the ladder's own
proportions. ``highlight-low`` is the hairline ``edge`` and ``highlight-med``
the active ``edge-hi`` — which is exactly what upstream says they are for
(cursor line, borders) — and the text ladder is ``text`` → ``subtle`` →
``muted``, with ``dim`` interpolated between the last two because this UI
has one more text rung than Rosé Pine does.

The accent mapping is the one editorial decision, and it follows upstream's
own usage table rather than hue:

- ``love`` is the documented terminal red and error colour → ``danger``.
- ``gold`` is the documented strings-and-warnings colour → ``warning`` and
  ``string``, both at once, which is precisely what upstream assigns it.
- ``foam`` is the documented "object keys, info, git add" hue → ``signal``,
  this UI's link/file/reference role.
- ``iris`` is the documented links-and-hints magenta → ``label``.
- ``pine`` is the documented terminal green (functions, git rename) →
  ``success``. It is a BLUE-green, not a leaf green, which is Rosé Pine's
  signature inversion and is kept.
- ``rose`` is left for ``accent`` — the theme's own name, and the one hue
  upstream spends on "git change, git dirty": the colour of something
  happening. It carries the caret, the live indicator and focus here.

Contrast is where the variants diverge, and each deviation below carries the
canonical value and its measured ratio. Main and moon needed two lifts each;
dawn is a near-total re-solve, for the reason ``lights.py`` documents — a hue
tuned to glow on a near-black ground arrives at 2–3:1 on paper.
"""

from __future__ import annotations

from local_operator.tui.theme import ThemeSpec

PALETTES: list[ThemeSpec] = [
    # Rosé Pine (main) — the original night. Every neutral is canonical
    # except the two this UI adds: `overlay` (a fourth elevation step past
    # canonical overlay #26233a, in the ladder's own ~1.1x proportion) and
    # `sunken` (one step below base, where the status band sits).
    #
    # Two accents needed work against the 4:1 state floor:
    #   - pine #31748f measures 3.38:1 on base and 3.16:1 on surface — the
    #     ONLY canonical accent under the floor on this ground. It lifts to
    #     #3a8aaa: same blue-green, 4.54:1 / 4.24:1.
    #   - muted #6e6a86 is canonical `faint` here (3.42:1) rather than a
    #     text rung, so `dim` is the subtle/muted midpoint #7f7b98 (4.36:1)
    #     — the extra rung this UI has and Rosé Pine does not.
    ThemeSpec(
        name="rose-pine",
        label="Rosé Pine",
        description="All natural pine, faux fur and a bit of soho vibes",
        dark=True,
        tokens={
            "bg": "#191724",  # canonical base
            "surface": "#1f1d2e",  # canonical surface
            "raised": "#26233a",  # canonical overlay
            "overlay": "#322f45",  # one step past canonical overlay (no upstream value)
            "sunken": "#13111c",  # one step below base (no upstream value)
            "fg": "#e0def4",  # canonical text
            "muted": "#908caa",  # canonical subtle
            "dim": "#7f7b98",  # subtle/muted midpoint — this UI's extra rung
            "faint": "#6e6a86",  # canonical muted (3.42:1, the inert rung)
            "edge": "#21202e",  # canonical highlight-low
            "edge-hi": "#403d52",  # canonical highlight-med
            "accent": "#ebbcba",  # rose — the theme's name and its "something changed" hue
            "success": "#3a8aaa",  # pine #31748f: 3.38:1 on base (< 4)
            "warning": "#f6c177",  # gold — upstream's warnings colour
            "danger": "#eb6f92",  # love — upstream's error/terminal-red
            "string": "#f6c177",  # gold — upstream assigns strings to gold too
            "signal": "#9ccfd8",  # foam — upstream's info/object-key hue
            "label": "#c4a7e7",  # iris — upstream's links/hints magenta
            "tint-danger": "#2e1c28",  # love's cast at base luminance (love reads 5.50:1)
            "tint-select": "#2a2033",  # rose cast — selection wears the accent
            "tint-select-hi": "#342839",
            "tint-attach": "#1e2b39",  # foam-leaning, so `signal` reads on the chip
            "tint-attach-hi": "#2d4457",
        },
    ),
    # Rosé Pine Moon — the same palette one shade warmer and lighter, and
    # the variant most people mean by "Rosé Pine". Grounds are canonical
    # (base #232136, surface #2a273f, overlay #393552 as edge-hi's partner);
    # `raised` is interpolated between surface and canonical overlay because
    # moon's own overlay is spent on the fourth step.
    #
    # The lighter ground costs contrast on exactly three tokens:
    #   - subtle #908caa drops to 4.46:1 on the surface step (< 4.5), so
    #     `muted` is #938fad — the same violet-gray, one step up.
    #   - pine #3e8fb0 measures 3.94:1 on surface (< 4) and lifts to #4195b7.
    #   - rose #ea9a97 is only 43 RGB units from love #eb6f92 on this
    #     variant (main's are 87 apart), close enough that the caret and an
    #     error message read as one hue. It desaturates to #eaaca9 — still
    #     moon's rose, 65 units off love.
    ThemeSpec(
        name="rose-pine-moon",
        label="Rosé Pine Moon",
        description="Rosé Pine's warmer moonlit ground, the soho night one shade up",
        dark=True,
        tokens={
            "bg": "#232136",  # canonical base
            "surface": "#2a273f",  # canonical surface
            "raised": "#312e4a",  # surface->overlay midpoint (no upstream value)
            "overlay": "#393552",  # canonical overlay
            "sunken": "#1c1a2c",  # one step below base (no upstream value)
            "fg": "#e0def4",  # canonical text
            "muted": "#938fad",  # canonical subtle #908caa: 4.46:1 on surface (< 4.5)
            "dim": "#7f7b98",  # subtle/muted midpoint — this UI's extra rung
            "faint": "#6e6a86",  # canonical muted
            "edge": "#2a283e",  # canonical highlight-low
            "edge-hi": "#44415a",  # canonical highlight-med
            "accent": "#eaaca9",  # rose #ea9a97: 43 RGB units from love — too close
            "success": "#4195b7",  # pine #3e8fb0: 3.94:1 on surface (< 4)
            "warning": "#f6c177",  # gold
            "danger": "#eb6f92",  # love
            "string": "#f6c177",  # gold
            "signal": "#9ccfd8",  # foam
            "label": "#c4a7e7",  # iris
            "tint-danger": "#392334",
            "tint-select": "#332a44",
            "tint-select-hi": "#3d3350",
            "tint-attach": "#25334a",
            "tint-attach-hi": "#374c6b",
        },
    ),
    # Rosé Pine Dawn — the light variant, and a near-total re-solve of the
    # accent row for the reason `lights.py` states: a hue that glows on a
    # near-black ground lands at 2–3:1 on paper. Every canonical dawn accent
    # except pine misses the 4:1 state floor (gold 2.05:1, rose 2.60:1,
    # foam 3.14:1, iris 3.47:1, love 3.84:1), so each is darkened along its
    # own hue by the SMALLEST step that clears 4:1 on both grounds — the
    # deviations are 0–25 ΔE, listed per token below: pine needs none, gold
    # (the furthest) needs 25.2.
    #
    # The ground ladder also has to invert. Canonical dawn `surface`
    # #fffaf3 is LIGHTER than `base` #faf4ed, so it cannot be this UI's
    # first elevation step (light polarity darkens upward); it becomes
    # `sunken`, the band below the paper, which is what a lighter-than-page
    # value is actually good for. The ladder then steps down through
    # highlight-low #f4ede8 and canonical overlay #f2e9e1.
    #
    # One pairing stays deliberately close: love and rose both collapse
    # toward the same dusty red under the floor (best achievable separation
    # is ~12 ΔE at full fidelity). `accent` takes rose at 16.8 ΔE from
    # canonical and `danger` takes love at 5.0 ΔE, buying 29 RGB units of
    # separation — tighter than the dark variants, in line with the closest
    # pairings the other light palettes here already ship.
    ThemeSpec(
        name="rose-pine-dawn",
        label="Rosé Pine Dawn",
        description="Rosé Pine at sunrise, warm paper under a soho morning",
        dark=False,
        tokens={
            "bg": "#faf4ed",  # canonical base
            "surface": "#f4ede8",  # canonical highlight-low
            "raised": "#f2e9e1",  # canonical overlay
            "overlay": "#e8ddd4",  # one step past overlay (no upstream value)
            "sunken": "#fffaf3",  # canonical surface — lighter than base, so it sinks
            "fg": "#4d496b",  # canonical text #575279: 6.66:1 on base (< 7)
            "muted": "#6a6681",  # canonical subtle #797593: 4.02:1 (< 4.5)
            "dim": "#7e7a89",  # canonical muted #9893a5: 2.73:1 (< 3.4)
            "faint": "#a5a0af",  # near canonical muted — the inert rung
            "edge": "#eee5dc",  # between highlight-low and highlight-med
            "edge-hi": "#dfdad9",  # canonical highlight-med
            "accent": "#a8594f",  # rose #d7827e: 2.60:1 (< 4) — 16.8 ΔE, held off love
            "success": "#286983",  # pine — CANONICAL (5.59:1/5.27:1), the only one that clears
            "warning": "#9e6200",  # gold #ea9d34: 2.05:1 (< 4) — the largest deviation
            "danger": "#a8566c",  # love #b4637a: 3.84:1 (< 4) — 5.0 ΔE
            "string": "#9e6200",  # gold, same solve as warning
            "signal": "#41787a",  # foam #56949f: 3.14:1 (< 4), held off pine
            "label": "#7d6694",  # iris #907aa9: 3.47:1 (< 4) — 7.8 ΔE
            "tint-danger": "#f6e2e3",
            # Selection/focus, AUTHORED rather than left on the rose cast.
            # `#f0e8e3` measured ΔE 1.71 from `surface` — a focused tool row
            # was indistinguishable from an unfocused one, and dawn was the
            # only theme with the defect (dark 8.54, github-light 8.74,
            # solarized-light 6.65). The paper has no well to sink into, so
            # the step is bought in CHROMA at near-constant luminance.
            #
            # IRIS: dawn's canonical meta accent, already what `label`
            # spends, which makes it the right hue for a selection that is
            # chrome rather than content. Measured: ΔE 10.05 from `surface`,
            # at 1.157:1 against `bg` — well inside the 2.2 tint ceiling.
            "tint-select": "#e9e2f2",
            "tint-select-hi": "#ded2ee",
            "tint-attach": "#e8eef0",  # foam-leaning wash
            "tint-attach-hi": "#d2e0e2",
            # The user's own prompt ground. The per-theme derivation solves
            # for maximum separation and lands on green-teal (159°) here,
            # which is correct by the metric and wrong for the eye: it fights
            # the warm paper. Blue is the one hue family dawn's palette does
            # not already spend — love/gold/rose are warm, pine/foam sit at
            # 180-200°, iris is violet — so a cool cast separates from the
            # ground instead of competing with it. Measured: ΔE 10.0 from
            # `bg` and 9.8 from its worst rival ground, both above the §9
            # floors, at 245° — the best of the blues tested (230/245/260).
            # Tool categories. The derivation leaves mutate/exec on the
            # neutral ramp, which is the safe default for 54 themes but keeps
            # dawn's ledger a single grey. Dawn has six solved accents and was
            # spending two, so these are authored from the canonical set it
            # already carries: foam for reading, gold for changing a file,
            # iris for running something, rose for coordination.
            #
            # Every value is a token dawn has already solved for WCAG on this
            # paper (all four measure >= 4.31:1 on `surface`), the four
            # separate from each other by >= 22.1 ΔE00, and each stays >= 11.6
            # ΔE00 from BOTH outcome hues — a category must never be mistaken
            # for a verdict. `success` is deliberately not reused here: it is
            # the ✓ one column to the right on the same row.
            "tool-read": "#41787a",  # foam — same hue as file paths, deliberately
            "tool-mutate": "#9e6200",  # gold
            "tool-exec": "#7d6694",  # iris
            "tool-meta": "#a8594f",  # rose
        },
    ),
]
